### ISSUE TITLE
[REVIEW]: OPEM (Open Source PEM Fuel Cell Simulation Tool)

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#676]
**Submitting author:** @sepandhaghighi (<a href="http://orcid.org/0000-0001-9450-2375">Sepand Haghighi</a>)
**Repository:** <a href="https://github.com/ECSIM/opem" target ="_blank">https://github.com/ECSIM/opem</a>
**Version:** v0.8
**Editor:** @katyhuff
**Reviewer:** @nnadeau
**Archive:** <a href="https://doi.org/10.5281/zenodo.1311905" target="_blank">10.5281/zenodo.1311905</a>

## Status

[![status](http://joss.theoj.org/papers/d1fa8d2a48cb2e723dc6d16797e463d0/status.svg)](http://joss.theoj.org/papers/d1fa8d2a48cb2e723dc6d16797e463d0)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/d1fa8d2a48cb2e723dc6d16797e463d0"><img src="http://joss.theoj.org/papers/d1fa8d2a48cb2e723dc6d16797e463d0/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/d1fa8d2a48cb2e723dc6d16797e463d0/status.svg)](http://joss.theoj.org/papers/d1fa8d2a48cb2e723dc6d16797e463d0)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/ECSIM/opem" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@nnadeau, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @katyhuff know.

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/ECSIM/opem">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v0.8)?
- [x] **Authorship:** Has the submitting author (@sepandhaghighi) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon. I'm here to help you with some common editorial tasks. @nnadeau it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/joss-papers/blob/joss.00676/joss.00676/10.21105.joss.00676.pdf)

**Comment by nnadeau** [comment]
Hi @sepandhaghighi and @katyhuff, 

After performing an initial review, I have submitted the following issues:

## Plagarism
Unfortunately, https://github.com/ECSIM/opem/issues/47 appears to be a case of plagiarism in the documentation. Considering that the entirety of `PEM.md` is copied, permission should be granted from the original author. Otherwise the reader should be referred directly to the original source.

## Paper
- https://github.com/ECSIM/opem/issues/46

## General syntax/grammer
The following issues reduce the readability of the documentation, examples, and paper:
- https://github.com/ECSIM/opem/issues/45
- https://github.com/ECSIM/opem/issues/44
- https://github.com/ECSIM/opem/issues/43
- https://github.com/ECSIM/opem/issues/42
- https://github.com/ECSIM/opem/issues/35

## Code Quality
The following issues describe desired improvements to the codebase:
- https://github.com/ECSIM/opem/issues/41
- https://github.com/ECSIM/opem/issues/40
- https://github.com/ECSIM/opem/issues/39
- https://github.com/ECSIM/opem/issues/38
- https://github.com/ECSIM/opem/issues/37
- https://github.com/ECSIM/opem/issues/36
- https://github.com/ECSIM/opem/issues/34

**Comment by sepandhaghighi** [comment]
Hi @nnadeau 
Thank you for your time
The issues have been fixed. ;-)

**Comment by katyhuff** [comment]
Thank you for this detailed review @nnadeau . 

@sepandhaghighi : all of the issues remain open. Please bring them to closure through conversation with @nnadeau and I. I have left comments on most regarding whether I believe the issues have been sufficiently handled or whether I think the issue looks like it has truly been fixed. 

Thanks for working quickly on these issues @sepandhaghighi . I am particularly concerned about the one regarding pem.md. The initial response by your team doesn't seem to recognize the gravity of plaigarism. I hope it will be handled well and swiftly.

**Comment by nnadeau** [comment]
@katyhuff @sepandhaghighi I'm still in the process of reviewing the actual codebase.

See https://github.com/ECSIM/opem/issues/50 for my initial observations regarding testing.

**Comment by sepandhaghighi** [comment]
@nnadeau @katyhuff 

I merged `develop` and `master` branches and most of issues with `fixed` label closed automatically.

Thanks

**Comment by nnadeau** [comment]
@katyhuff @sepandhaghighi 

My final review and opinion:

While the package has an `obvious research application`, it does not provide `a significant contribution to the available open source software that either enables some new research challenges to be addressed or makes addressing research challenges significantly better (e.g., faster, easier, simpler)`. (https://joss.theoj.org/about#author_guidelines)

As the majority of function are simple one-liner multiplications, from my perspective, this is more of a `minor, 'utility' package` that would most likely be re-coded by most researchers using `NumPy` and `SciPy`. The most "novel" components are the pretty-print style outputs.

Moreover, the lack of good coding and Python practice makes this difficult for the open-source and research communities to extend and build upon this work. In its current form, this package is best suited for use in a single lab and research group.

@sepandhaghighi, while my overall review may be disappointing, I am very impressed by your team's effort and amazing documentation/README/website. I am also very pleased to see scientific contributions to the Python community. However, from a  journal perspective, I cannot give my stamp of approval. 

Best of luck with your future research!

**Comment by sepandhaghighi** [comment]
@nnadeau 

Thanks for your time and deep review
Your points about code style were very good and detailed and we will solve all of them.
But about usage of this package really is not like this, we decided to create this package from some real problems in `PEM` research and focus on models to collect them in one package.
OPEM designed for `PEM` researchers without programming knowledge, it is simple because we collect and solved equations outside of code.
We spent about 10 months to collect and validate these models, now is really easy to re-code but this procedure is not.
There is no package like OPEM that support all of these models in this style and I think we provide easier and simpler way in `PEM` analysis and we are in `JOSS` scope.

Thanks ;-)

**Comment by sepandhaghighi** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/joss-papers/blob/joss.00676/joss.00676/10.21105.joss.00676.pdf)

**Comment by arfon** [comment]
@katyhuff - I think this review is waiting on your input.

**Comment by katyhuff** [comment]
Hi Folks, I'm sorry for the delay here. It took some time and exploration to determine whether there is really a fit concern for JOSS. @nnadeau , I thank you for your detailed and careful review. I understand your concern about this potentially being a minor utility rather than a tool enabling new research. However, after investigating a bit, I personally evaluate the work more generously. While it may only enable new research in this small subdomain, and perhaps only in this research group, I am convinced that it is more than a minor utility and will indeed, enable new research. I would like to approve this paper.


@sepandhaghighi @kasraaskari At this point could you make an archive of the reviewed software in Zenodo/figshare/other service and update this thread with the DOI of the archive? I can then move forward with accepting the submission.

**Comment by sepandhaghighi** [comment]
@katyhuff Thanks for your time.

Reviewed software  DOI : 10.5281/zenodo.1311905

**Comment by katyhuff** [comment]
@whedon set 10.5281/zenodo.1311905 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.1311905" target="_blank">10.5281/zenodo.1311905</a> is the archive.

**Comment by katyhuff** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[--> Check article proof :page_facing_up: <--](https://github.com/openjournals/joss-papers/blob/joss.00676/joss.00676/10.21105.joss.00676.pdf)

**Comment by katyhuff** [comment]
Thank you @nnadeau for reviewing this. Thank you @sepandhaghighi and @kasraaskari for the submission! 

@arfon we're ready to accept this!

**Comment by arfon** [comment]
@nnadeau - many thanks for your review here and to @katyhuff for editing this submission ✨ 

@sepandhaghighi - your paper is now accepted into JOSS and your DOI is https://doi.org/10.21105/joss.00676 :zap: :rocket: :boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](http://joss.theoj.org/papers/10.21105/joss.00676/status.svg)](https://doi.org/10.21105/joss.00676)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.00676">
  <img src="http://joss.theoj.org/papers/10.21105/joss.00676/status.svg" alt="DOI badge" >
</a>
```
This is how it will look in your documentation:

[![DOI](http://joss.theoj.org/papers/10.21105/joss.00676/status.svg)](https://doi.org/10.21105/joss.00676)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: http://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://www.flipcause.com/secure/cause_pdetails/Mjk3Nzk=

**Comment by sepandhaghighi** [comment]
Thanks @arfon @katyhuff @nnadeau  @whedon !
